internal/eventstore/eventstore.go

```diff
@@ -100,12 +106,24 @@ func (es *Eventstore) Push(ctx context.Context, cmds ...Command) ([]Event, error
 	// https://github.com/zitadel/zitadel/issues/7202
 retry:
 	for i := 0; i <= es.maxRetries; i++ {
-		events, err = es.pusher.Push(ctx, cmds...)
+		events, err = es.pusher.Push(ctx, client, cmds...)
+		// if there is a transaction passed the calling function needs to retry
+		if _, ok := client.(database.Tx); ok {
+			break retry
+		}
 		var pgErr *pgconn.PgError
-		if !errors.As(err, &pgErr) || pgErr.ConstraintName != "events2_pkey" || pgErr.SQLState() != "23505" {
+		if !errors.As(err, &pgErr) {
 			break retry
 		}
-		logging.WithError(err).Info("eventstore push retry")
+		if pgErr.ConstraintName == "events2_pkey" && pgErr.SQLState() == "23505" {
+			logging.WithError(err).Info("eventstore push retry")
```

Comment:
Would it be helpful to add prom metrics for here 23505 and 40001? Regarding 40001 in particular - a high rate of serialization errors might be an indicator that the serializable isolation level isn't performing well.

Reply:
thanks for your comment, we will merge https://github.com/zitadel/zitadel/pull/8816 as well which changes this logic anyways. But I agree that this would be a useful metric. I created this issue as a follow up https://github.com/zitadel/zitadel/issues/8948